pkg/lib/concurrency/channel_transformer.go

```diff
@@ -0,0 +1,41 @@
+package concurrency
+
+import "context"
+
+// AsyncChannelTransform copies messages from an input channel, transforms them, and sends them to an output channel.
+func AsyncChannelTransform[In any, Out any](
+	ctx context.Context,
+	input <-chan In,
+	bufferCapacity int,
+	transform func(In) (*AsyncResult[Out], error)) <-chan *AsyncResult[Out] {
+	output := make(chan *AsyncResult[Out], bufferCapacity)
+
+	go func() {
+		defer close(output)
+		for {
+			select {
+			case msg, ok := <-input:
+				if !ok {
+					return // Input channel closed
+				}
+				result := new(AsyncResult[Out])
+				transformedValue, err := transform(msg)
+				if err != nil {
+					result.Err = err
+				} else {
+					result = transformedValue
+				}
```

Comment:
If the `transform` function returned an Out and an error then this code could be...

```suggestion
				result := NewAsyncResult[Out](transform(msg))
```

Reply:
Wow .. Nice!